azure-proto-core/Resources/Plan.cs

```diff
@@ -17,12 +16,43 @@ public class Plan : IEquatable<Plan>, IComparable<Plan>
 
         public int CompareTo(Plan other)
         {
-            throw new NotImplementedException();
+            if (other == null) return 1;
+            if (this.Name.CompareTo(other.Name) == 0)
+            {
+                if (this.Product.CompareTo(other.Product) == 0)
+                {
+                    if (this.PromotionCode.CompareTo(other.PromotionCode) == 0)
+                    {
+                        if (this.Publisher.CompareTo(other.Publisher) == 0)
+                        {
+                            return this.Version.CompareTo(other.Version);
+                        }
+                        return this.Publisher.CompareTo(other.Publisher);
+                    }
+                    return this.PromotionCode.CompareTo(other.PromotionCode);
+                }
+                return this.Product.CompareTo(other.Product);
+            }
+            return this.Name.CompareTo(other.Name);
         }
 
         public bool Equals(Plan other)
         {
-            throw new NotImplementedException();
+            if (other == null) return false;
+            if (this.Name.Equals(other.Name))
+            {
```

Comment:
Why the super nested if checks here? Can add these more into the one line.

Reply:
Here comparing in this sequence: Name > Product > PromotionCode > Publisher > Version. For example, when name's value is not same, will skip other properties comparisons.

Reply:
&& has the same effect, no?

Reply:
If we don't handle the case when a property value is null, we can use &&. Now the comparison is more complex, so I have updated to without using nesting if statements.